Enable embedding images by URL

adamsilverstein opened the pull request:
Addresses https://github.com/xwp/wp-core-media-widgets/issues/9

* Add a custom Post modal state to handle our specific needs (select/embed)
* Modify the insert button language temporarily - cleanest approach for insert button. I tried overriding the setup methods and that was messier. 
* restore standard post state and language on close
* handle embed data on close
* Extend the template to handle embedded images

Q: Not sure about the edit dialog - or the edit details sidebar in the media select mode: do we really want all of these attributes editable? Maybe the widget itself has a title and description and the media part is only image selection? I'm confused about how these would be used - related: https://core.trac.wordpress.org/ticket/32417#comment:132

![](https://cl.ly/2W2V451n3a1x/Screen%20Recording%202017-03-25%20at%2010.58%20PM.gif)

melchoyce commented:
Looking good!

Some things I noticed:

If you go back and edit an image after picking it for the widget (for example, cropping it, or flipping it horizontally), the updated image appears in your site preview, but not in the widget preview:

<img width="571" alt="screen shot 2017-03-27 at 1 45 40 pm" src="https://cloud.githubusercontent.com/assets/2846578/24369945/a808a88a-12f3-11e7-95b3-5bd13d7233c8.png">

Also, when you change images, the image settings (size, link, etc.) revert to default. I just noticed that you also can't edit these when selecting or changing an image, only when editing. You should be able to change these settings while you're picking an image, IMO.

obenland commented:
When updating an URL-image I get an undefined error  for [`mediaFrameContentView.model.dfd`](https://github.com/xwp/wp-core-media-widgets/blob/c47c013e206726d79b9024d320b56022e27bc0e6/wp-admin/js/widgets/media-image-widget.js#L148)

adamsilverstein commented:
@melchoyce @obenland  - thanks for the review and feedback, I'll work on addressing the issues you mentioned.

adamsilverstein commented:
@obenland I addressed the console error you were seeing in https://github.com/xwp/wp-core-media-widgets/pull/46/commits/047e7e62a2a5a6b8cff930498b1380fb4c929043

adamsilverstein commented:
With the latest commits, I have completed support for embedding remote images via URL. This should be close to or ready for merge, the basic functionality is in place.

In my testing, all of the fields including the caption, alignment, and the advanced link and class options all work correctly for embedded by url media.

These fields:

![](https://cl.ly/1u0K1R051e1E/Customize_New_post__developwordpress_2017-04-02_11-42-40.jpg)

all render correctly:

![](https://cl.ly/3i22193a2Z0U/New_post__developwordpress_2017-04-02_11-43-53.jpg)

I'd like to address the issues [@melchoyce raised](https://github.com/xwp/wp-core-media-widgets/pull/46#issuecomment-289530257) on a separate issue.

westonruter commented on wp-admin/js/widgets/media-image-widget.js:
```diff
@@ -43,33 +43,28 @@
+					image_title: '',
+					link_classes: '',
+					link_rel: '',
+					link_url: attachment.link,
```
Shouldn't this be `attachment.linkUrl`?

westonruter commented on wp-admin/js/widgets/media-image-widget.js:
```diff
@@ -147,11 +142,17 @@
+				_.defer( function() {
+					jQuery( '.media-frame-content [data-setting="caption"]:first' ).focus();
+				} );
+			}
```
We could also just remove all of this autofocus logic from the widget itself. To me it seems like this should actually be handled in the media library JS itself, right? I mean, right now it seems that the [`this.$el.focus();` call](https://github.com/WordPress/wordpress-develop/blob/4.7.3/src/wp-includes/js/media-views.js#L6805) in `Modal#focus()` should perhaps be improved to automatically autofocus on the first frame's first focusable element?

adamsilverstein replied:
I'm happy to remove this logic here and agree the media modal should be taking care of the focus. It does this already focusing on the modal itself, which may be enough.

westonruter commented on wp-admin/js/widgets/media-widgets.js:
```diff
@@ -1,5 +1,5 @@
 /* eslint consistent-this: [ "error", "control" ] */
-wp.mediaWidgets = ( function( $ ) {
+wp.mediaWidgets = ( function( $, _wpMediaViewsL10n ) {
```
Instead of accessing this “private” global variable, should we not instead access `wp.media.view.l10n`?

westonruter commented on wp-admin/js/widgets/media-widgets.js:
```diff
@@ -141,6 +236,21 @@ wp.mediaWidgets = ( function( $ ) {
+				attachment = new wp.media.model.Attachment( control.model.attributes );
+
+				// Once the widget renders, inject the preview.
+				_.defer( function() {
```
This `defer` seems a bit brittle and it may indicate we should improve something with how `selectedAttachment` is used.

adamsilverstein replied:
Couldn't quite track down why this is required - without it, the embedded items are set immediately and never rendered. It may be the sequence of event handlers and how they are added. We can dig into this more post merge.

westonruter replied:
Fixed in 0b84597

westonruter commented on wp-admin/js/widgets/media-widgets.js:
```diff
@@ -131,6 +132,100 @@ wp.mediaWidgets = ( function( $ ) {
+			 */
+			control.originalButtonLanguage = _wpMediaViewsL10n.insertIntoPost;
+			control.originalMediaFramePost = wp.media.view.MediaFrame.Post;
+			control.customMediaFramePost = wp.media.view.MediaFrame.Post.extend( {
```
I think this should be `CustomMediaFramePost` since it is a class.

westonruter commented on wp-includes/widgets/class-wp-widget-image.php:
```diff
@@ -148,49 +148,69 @@ public function render_media( $instance ) {
+					'[caption id="0" align="%s" width="%d"]<img class="" src="%s" alt="%s" width="%d" height="%d" />%s[/caption]',
+					$align,
+					$width,
+					$url,
```
Add `esc_url()` for good measure?

westonruter commented on wp-includes/widgets/class-wp-widget-image.php:
```diff
@@ -148,49 +148,69 @@ public function render_media( $instance ) {
+					$align,
+					$width,
+					$url,
+					$alt,
```
Add `esc_attr()`?

westonruter commented on wp-includes/widgets/class-wp-widget-image.php:
```diff
@@ -148,49 +148,69 @@ public function render_media( $instance ) {
+			$image = do_shortcode(
+				sprintf(
+					'[caption id="0" align="%s" width="%d"]<img class="" src="%s" alt="%s" width="%d" height="%d" />%s[/caption]',
+					$align,
```
Add `esc_attr()` for crossing i's and dotting t's?

westonruter commented on wp-includes/widgets/class-wp-widget-image.php:
```diff
@@ -148,49 +148,69 @@ public function render_media( $instance ) {
+			$height  = isset( $instance['height'] ) ? $instance['height'] : '';
+			$image = do_shortcode(
+				sprintf(
+					'[caption id="0" align="%s" width="%d"]<img class="" src="%s" alt="%s" width="%d" height="%d" />%s[/caption]',
```
Why not just call `img_caption_shortcode()` directly?

adamsilverstein commented:
@westonruter thanks for the review and feedback!

obenland commented:
> Q: Not sure about the edit dialog - or the edit details sidebar in the media select mode: do we really want all of these attributes editable?

I'm not sure how I feel about them being removed. Without the controls it looks like any attached image gets automatically liked to its attachment page. It's also inconsistent with the embed modal where it still lets you align images and add a custom link url.
While I understand that aligning an image within a widget probably doesn't make much sense, the other attributes feel like they still have a place there.

@westonruter With the exception of https://github.com/xwp/wp-core-media-widgets/pull/46#discussion_r109343180 and https://github.com/xwp/wp-core-media-widgets/pull/46#discussion_r109343522 your feedback should be addressed.

obenland commented:
One bug that I noticed: When pasting a non-image url, like an audio file, it returns the most recently uploaded picture

adamsilverstein commented:
@obenland thanks for your work here - i can look at the remaining issues.

westonruter commented on wp-admin/js/widgets/media-image-widget.js:
```diff
@@ -69,6 +88,38 @@
+					image_title: '',
+					link_classes: '',
+					link_rel: '',
+					link_url: attachment.linkUrl ? attachment.linkUrl : attachment.link,
```
Is `attachment.link` a URL when `attachment.linkUrl` is empty?

adamsilverstein replied:
I thought link was the link type (self, attachment page, other) and the linkUrl was the actual URL. I will test this out.

adamsilverstein replied:
here is the passed data when I edit an embedded image:

![](https://d3vv6lp55qjaqc.cloudfront.net/items/1T2v2m0m1G2F0Y0c413Q/Image%202017-04-04%20at%208.20.21%20AM.png?X-CloudApp-Visitor-Id=1371633)

westonruter replied:
Yes, `link` is the type and `linkUrl` is the actual URL. So I don't think that the type (`attachment.link`) should ever be made the `link_url`.

obenland commented on wp-admin/js/widgets/media-image-widget.js:
```diff
@@ -106,7 +106,7 @@
 					link_classes: '',
 					link_rel: '',
-					link_url: attachment.linkUrl ? attachment.linkUrl : attachment.link,
+					link_url: attachment.linkUrl,
```
`linkUrl` is only set when it's a custom link. When it's linked to the image, it doesn't pass the image url again.
So it looks like we need to be more verbose here and check for what's in `link` and populate `link_url` accordingly.

westonruter replied:
¿So like:

```js
link_url: 'custom' === attachment.link ? attachment.linkUrl : '',
```

obenland replied:
I currently have 
```javascript
if ( 'custom' === attachment.link ) {
	linkUrl = attachment.linkUrl;
} else if ( 'file' === attachment.link ) {
	linkUrl = attachment.url;
}
```
on my sandbox.

westonruter replied:
@obenland that should work, but it would seem to be somewhat redundant. If the `link_type` is `file` then the PHP logic isn't going to look at `link_url` anyway, and will instead look at `url` anyway, per `\WP_Widget_Image::render_media()`:

```php
$url = '';
if ( 'file' === $instance['link_type'] ) {
	$url = $attachment ? wp_get_attachment_url( $attachment->ID ) : $instance['url'];
} elseif ( $attachment && 'post' === $instance['link_type'] ) {
	$url = get_attachment_link( $attachment->ID );
} elseif ( 'custom' === $instance['link_type'] && ! empty( $instance['link_url'] ) ) {
	$url = $instance['link_url'];
}
```

westonruter commented on wp-admin/js/widgets/media-image-widget.js:
```diff
@@ -187,23 +187,6 @@
-				_.defer( function() {
-					jQuery( '.media-frame-content [data-setting="caption"]:first' ).focus();
-				} );
-			}
```
@adamsilverstein do you think that the improvements to media modal focusing will be made to the underlying media JS? Do you concur that there seems to be a problem with the focus behavior when opening the modal?

obenland commented on wp-admin/js/widgets/media-image-widget.js:
```diff
@@ -98,7 +98,7 @@
 			if ( ! _.isEmpty( attachment ) ) {
 				props = {
-					attachment_id: attachment.id,
+					attachment_id: 0,
```
This causes my control preview to not update some times. But not always. Still trying to work out why. Does this work for everyone else?

westonruter replied:
I think this may be fixed as of 8c3ae25

westonruter commented on wp-admin/js/widgets/media-widgets.js:
```diff
@@ -397,6 +397,7 @@ wp.mediaWidgets = ( function( $ ) {
 
 				// Update widget instance.
 				control.model.set( control.getSelectFrameProps( mediaFrame ) );
+				control.model.trigger( 'update' );
```
Why a custom `update` event here? Wouldn't it suffice to just change the above `change:attachment_id` to just `change` and then this custom event trigger could be removed?

westonruter replied:
The `update` event is supposed to be for collections:
> single event triggered after any number of models have been added or removed from a collection.

westonruter replied:
@adamsilverstein changed in 8c3ae2581cf580

westonruter commented on wp-admin/js/widgets/media-image-widget.js:
```diff
@@ -43,32 +43,78 @@
+
+			if ( ! _.isEmpty( attachment ) ) {
+				props = {
+					attachment_id: 0,
```
Why is this `attachment_id: 0` and not `attachment_id: attachment.id`?

westonruter replied:
I see. The `attachment.id` is retained in `ImageWidgetModel#_getAttachmentProps` but is not relevant here in `ImageWidgetModel#_getEmbedProps` since there is no underlying attachment.

westonruter commented:
@melchoyce the changes here remove the Attachment Display Settings from the select frame:

![image](https://cloud.githubusercontent.com/assets/134745/24673536/89c770c8-192d-11e7-8e0e-5e2f3acb3035.png)

Is that OK?

melchoyce commented:
Why are we removing attachments? Being able to link your image somewhere seems very useful.

westonruter commented on wp-admin/js/widgets/media-widgets.js:
```diff
@@ -258,95 +368,36 @@ wp.mediaWidgets = ( function( $ ) {
 
 				// Update cached attachment object to avoid having to re-fetch. This also triggers re-rendering of preview.
-				attachment = mediaFrame.state().get( 'selection' ).first().toJSON();
+				attachment = 'embed' === state.get( 'id' ) ? state.props.toJSON() : state.get( 'selection' ).first().toJSON();
```
I was getting as JS error here, though I'm trying to figure out how to reproduce:

![image](https://cloud.githubusercontent.com/assets/134745/24674879/fb0c4840-1931-11e7-8e3b-c17a9519762d.png)

westonruter replied:
It happens if I open the media modal and then close it without having selected anything.

westonruter replied:
It's because the event was changed from `select` to `close`. Working on a fix.

westonruter replied:
Fixed in d284087.

westonruter commented:
> Why are we removing attachments? Being able to link your image somewhere seems very useful.

@melchoyce attachments aren't being removed. It's just the Attachment _Display Settings_ fields on the select frame.

melchoyce commented:
But then how would you add a link?

melchoyce commented:
Testing — can't figure out how to add a link as I'm adding an image. Also noticed we're defaulting to thumbnail, which crops the image:

<img width="669" alt="screen shot 2017-04-04 at 3 41 46 pm" src="https://cloud.githubusercontent.com/assets/2846578/24675458/3b084172-194d-11e7-860c-742845b3d2a9.png">

It's confusing to see one thing in the widget and another in the preview. They should match. 

Can we default to a not-cropped version?

melchoyce commented:
After testing this against the master branch right now, I definitely prefer the defaults in the master branch.

adamsilverstein commented:
Did we loose the image settings from the regular image attachment frame as well? if so that is a mistake and I will fix that.

When selecting an embed image, you can't select an image size because we aren't importing and resizing the image, we are using it as is. i think you can add a link as you add the image?

adamsilverstein commented:
I see, linking is available from the 'insert from url' flow, but it got lost on the insert from media state. i will fix this @melchoyce and test against master again, I agree it should be available while you are adding the image.

![Insert from URL](https://cl.ly/350K2O37010W/Widgets__developwordpress__WordPress_2017-04-04_17-09-38.jpg)

westonruter commented:
> I see, linking is available from the 'insert from url' flow, but it got lost on the insert from media state. i will fix this @melchoyce and test against master again, I agree it should be available while you are adding the image.

@adamsilverstein @melchoyce I believe I've fixed this in https://github.com/xwp/wp-core-media-widgets/pull/46/files/0b84597..1ca00e5

adamsilverstein commented on wp-admin/js/widgets/media-widgets.js:
```diff
@@ -200,7 +204,7 @@ wp.mediaWidgets = ( function( $ ) {
 				 * @return {void}
 				 */
-				createStates: function() {
+				createStates: function createStates() {
```
@westonruter can you explain the advantages or why you prefer this pattern as opposed to 'anonymous' functions?

westonruter replied:
Yes. The reason for naming the functions is for stack traces. Compare:

![image](https://cloud.githubusercontent.com/assets/134745/24719317/234015f0-19ee-11e7-97b5-0525719c660f.png)

More reasons here: https://github.com/getify/You-Dont-Know-JS/blob/master/scope%20%26%20closures/ch3.md#anonymous-vs-named

adamsilverstein commented on wp-admin/js/widgets/media-widgets.js:
```diff
@@ -370,6 +367,7 @@ wp.mediaWidgets = ( function( $ ) {
 
 			if ( ! control.templateRendered ) {
 				control.$el.html( control.template()( control.model.attributes ) );
+				control.renderPreview(); // Hereafter it will re-render when control.selectedAttachment changes.
```
@westonruter Brilliant, thank you! I was struggling to find where this re-rendering was failing, this makes way more sense.

adamsilverstein commented:
I did some testing with the latest code and everything worked well, going to see if I can address the cropping/rotation in the preview.

adamsilverstein commented:
@melchoyce I fixed the issue with image cropping/rotation not showing in the preview...

In https://github.com/xwp/wp-core-media-widgets/pull/46/commits/36fc72588a56e3339d83d061ac2d6dee1afe8fb5 I switched the preview template to using only the image 'url' property for display. WordPress changes this data point when you rotate or crop an image, causing the preview to re-render correctly. This simplifies the template and we can control the preview size with CSS.

melchoyce commented:
Awesome! Need me to test?

westonruter commented:
@melchoyce yes, please test and once you verify we'll merge.

melchoyce commented:
🎉
